### ISSUE TITLE
Segmentation Fault when importing Torch

### ISSUE THREAD

**Issue by BernardoGO** [pytorch/pytorch#4101]
I've just installed PyTorch using:

> sudo pip3 install http://download.pytorch.org/whl/cu90/torch-0.3.0.post4-cp36-cp36m-linux_x86_64.whl

The installation process was successful, however, when I try to import torch I get:

```
Python 3.6.3 (default, Oct  3 2017, 21:45:48) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
Segmentation fault (core dumped)
```

I'm running on Ubuntu 17.10 with CUDA 9.0, CUDNN 7.0 and NVIDIA 384.90.

**Comment by soumith** [comment]
@BernardoGO  could you help us with two additional commands and outputs:

```
$ gdb python
(gdb) r -c "import torch"
```

additionally: 

```
ldd $(dirname $(which python))/../lib/python3.6/site-packages/torch/_C*.so
```

**Comment by BernardoGO** [comment]
@soumith Thanks for responding quickly. Here are the outputs:

```
(gdb)  r -c "import torch"
Starting program: /usr/bin/python -c "import torch"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff344a700 (LWP 18605)]
[New Thread 0x7ffff2c49700 (LWP 18606)]
[New Thread 0x7ffff0448700 (LWP 18607)]
[New Thread 0x7fffedc47700 (LWP 18608)]
[New Thread 0x7fffeb446700 (LWP 18609)]
[New Thread 0x7fffe8c45700 (LWP 18610)]
[New Thread 0x7fffe6444700 (LWP 18611)]
[Thread 0x7ffff344a700 (LWP 18605) exited]
[Thread 0x7ffff2c49700 (LWP 18606) exited]
[Thread 0x7ffff0448700 (LWP 18607) exited]
[Thread 0x7fffeb446700 (LWP 18609) exited]
[Thread 0x7fffe8c45700 (LWP 18610) exited]
[Thread 0x7fffe6444700 (LWP 18611) exited]
[Thread 0x7fffedc47700 (LWP 18608) exited]
[Inferior 1 (process 18600) exited normally]

```


`ldd: /usr/bin/../lib/python3.6/site-packages/torch/_C*.so: No such file or directory
`

Furthermore, I've noticed that none of my python folders contains site-packages.
```
ls $(dirname $(which python))/../lib/python3.6/si*

> /usr/bin/../lib/python3.6/signal.py         
> /usr/bin/../lib/python3.6/site.py
> /usr/bin/../lib/python3.6/sitecustomize.py

```

**Comment by soumith** [comment]
hmmm that's weird.
I wonder why gdb passed smoothly.

Is this also segfaullting?
```
python -c "import torch"
```

Also, can you print:
```
python -c "import torch; print(torch)"
```

**Comment by BernardoGO** [comment]
Both segfaulted

```
bernardo@ber:~$ python -c "import torch"
Segmentation fault (core dumped)
bernardo@ber:~$ python -c "import torch; print(torch)"
Segmentation fault (core dumped)
```

**Comment by soumith** [comment]
`**scratches head**`

how about:

```
gdb python
r -c "import torch; print(torch)"
```

**Comment by BernardoGO** [comment]
Ok, now it seems that it crashed with gdb:

```
(gdb) r -c "import torch; print(torch)"
Starting program: /usr/bin/python -c "import torch; print(torch)"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff344a700 (LWP 26765)]
[New Thread 0x7ffff2c49700 (LWP 26766)]
[New Thread 0x7ffff0448700 (LWP 26767)]
[New Thread 0x7fffedc47700 (LWP 26768)]
[New Thread 0x7fffeb446700 (LWP 26770)]
[New Thread 0x7fffe8c45700 (LWP 26771)]
[New Thread 0x7fffe6444700 (LWP 26772)]

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x0000000000002360 in ?? ()
```

**Comment by soumith** [comment]
and what is the backtrace?

```
gdb python
r -c "import torch; print(torch)"
bt
```

**Comment by BernardoGO** [comment]
Just noticed, path in my machine is: `/usr/local/lib/python3.6/dist-packages/torch/_C.`

```
(gdb) bt
#0  0x0000000000002360 in ?? ()
#1  0x00007ffff7de5a6a in call_init (l=<optimized out>, argc=argc@entry=3, 
    argv=argv@entry=0x7fffffffde28, env=env@entry=0xb85860) at dl-init.c:72
#2  0x00007ffff7de5b7b in call_init (env=0xb85860, argv=0x7fffffffde28, 
    argc=3, l=<optimized out>) at dl-init.c:30
#3  _dl_init (main_map=main_map@entry=0xeb4270, argc=3, argv=0x7fffffffde28, 
    env=0xb85860) at dl-init.c:120
#4  0x00007ffff7deab86 in dl_open_worker (a=a@entry=0x7fffffffb840)
    at dl-open.c:575
#5  0x00007ffff6d88d64 in __GI__dl_catch_error (objname=0x7fffffffb830, 
    errstring=0x7fffffffb838, mallocedp=0x7fffffffb82f, 
    operate=0x7ffff7dea7a0 <dl_open_worker>, args=0x7fffffffb840)
    at dl-error-skeleton.c:198
#6  0x00007ffff7dea0d9 in _dl_open (
    file=0x7fffdf813a60 "/usr/local/lib/python3.6/dist-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so", mode=-2147483391, 
    caller_dlopen=0x58de76 <_PyImport_FindSharedFuncptr+390>, nsid=-2, 
    argc=<optimized out>, argv=<optimized out>, env=0xb85860) at dl-open.c:660
#7  0x00007ffff79b2ff6 in dlopen_doit (a=a@entry=0x7fffffffba50) at dlopen.c:66
#8  0x00007ffff6d88d64 in __GI__dl_catch_error (objname=0xb85d10, 
    errstring=0xb85d18, mallocedp=0xb85d08, 
    operate=0x7ffff79b2fa0 <dlopen_doit>, args=0x7fffffffba50)
    at dl-error-skeleton.c:198
---Type <return> to continue, or q <return> to quit---
#9  0x00007ffff79b3759 in _dlerror_run (
    operate=operate@entry=0x7ffff79b2fa0 <dlopen_doit>, 
    args=args@entry=0x7fffffffba50) at dlerror.c:163
#10 0x00007ffff79b3092 in __dlopen (file=<optimized out>, mode=<optimized out>)
    at dlopen.c:87
#11 0x000000000058de76 in _PyImport_FindSharedFuncptr ()
#12 0x0000000000576b18 in _PyImport_LoadDynamicModuleWithSpec ()
#13 0x00000000005743c5 in ?? ()
#14 0x00000000004c47ed in PyCFunction_Call ()
#15 0x0000000000557040 in _PyEval_EvalFrameDefault ()
#16 0x000000000054efc1 in ?? ()
#17 0x000000000054f24d in ?? ()
#18 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#19 0x000000000054e4c8 in ?? ()
#20 0x000000000054f4f6 in ?? ()
#21 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#22 0x000000000054e4c8 in ?? ()
#23 0x000000000054f4f6 in ?? ()
#24 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#25 0x000000000054e4c8 in ?? ()
#26 0x000000000054f4f6 in ?? ()
#27 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#28 0x000000000054e4c8 in ?? ()
---Type <return> to continue, or q <return> to quit---
#29 0x000000000054f4f6 in ?? ()
#30 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#31 0x000000000054e4c8 in ?? ()
#32 0x00000000005582c2 in _PyFunction_FastCallDict ()
#33 0x0000000000459981 in _PyObject_FastCallDict ()
#34 0x000000000045b034 in _PyObject_CallMethodIdObjArgs ()
#35 0x000000000057581a in PyImport_ImportModuleLevelObject ()
#36 0x0000000000556add in _PyEval_EvalFrameDefault ()
#37 0x000000000054efc1 in ?? ()
#38 0x000000000054ff73 in PyEval_EvalCode ()
#39 0x000000000054cfda in ?? ()
#40 0x00000000004c47ed in PyCFunction_Call ()
#41 0x0000000000557040 in _PyEval_EvalFrameDefault ()
#42 0x000000000054efc1 in ?? ()
#43 0x000000000054f24d in ?? ()
#44 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#45 0x000000000054e4c8 in ?? ()
#46 0x000000000054f4f6 in ?? ()
#47 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#48 0x000000000054e4c8 in ?? ()
#49 0x000000000054f4f6 in ?? ()
#50 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#51 0x000000000054e4c8 in ?? ()
---Type <return> to continue, or q <return> to quit---
#52 0x000000000054f4f6 in ?? ()
#53 0x0000000000553aaf in _PyEval_EvalFrameDefault ()
#54 0x000000000054e4c8 in ?? ()
#55 0x00000000005582c2 in _PyFunction_FastCallDict ()
#56 0x0000000000459981 in _PyObject_FastCallDict ()
#57 0x000000000045b034 in _PyObject_CallMethodIdObjArgs ()
#58 0x000000000057581a in PyImport_ImportModuleLevelObject ()
#59 0x0000000000556add in _PyEval_EvalFrameDefault ()
#60 0x000000000054efc1 in ?? ()
#61 0x000000000054ff73 in PyEval_EvalCode ()
#62 0x000000000042c37a in PyRun_SimpleStringFlags ()
#63 0x0000000000441178 in Py_Main ()
#64 0x0000000000421f64 in main ()

```

**Comment by soumith** [comment]
what is:

```
ldd /usr/local/lib/python3.6/dist-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so
```

**Comment by BernardoGO** [comment]
```
ldd /usr/local/lib/python3.6/dist-packages/torch/_C.cpython-36m-x86_64-linux-gnu.so

	linux-vdso.so.1 =>  (0x00007ffcce1f5000)
	libshm.so => /usr/local/lib/python3.6/dist-packages/torch/lib/libshm.so (0x00007f37cdcac000)
	libcudart.so.9.0 => /usr/local/lib/python3.6/dist-packages/torch/lib/libcudart.so.9.0 (0x00007f37cda3d000)
	libnvToolsExt.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libnvToolsExt.so.1 (0x00007f37cd833000)
	libcudnn.so.7 => /usr/local/lib/python3.6/dist-packages/torch/lib/libcudnn.so.7 (0x00007f37bc5c0000)
	libTH.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTH.so.1 (0x00007f37b9a91000)
	libTHS.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTHS.so.1 (0x00007f37b9860000)
	libTHNN.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTHNN.so.1 (0x00007f37b951d000)
	libATen.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libATen.so.1 (0x00007f37b8d0f000)
	libTHC.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTHC.so.1 (0x00007f37a4353000)
	libTHCS.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTHCS.so.1 (0x00007f37a3fc3000)
	libTHCUNN.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libTHCUNN.so.1 (0x00007f379e8ef000)
	libnccl.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libnccl.so.1 (0x00007f379be65000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f379bc4e000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f379ba2f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f379b64f000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f37cf854000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f379b447000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f379b0f1000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f379aeed000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f379ab65000)
	libgomp.so.1 => /usr/local/lib/python3.6/dist-packages/torch/lib/libgomp.so.1 (0x00007f379a94e000)
	libcublas.so.9.0 => /usr/local/lib/python3.6/dist-packages/torch/lib/libcublas.so.9.0 (0x00007f3797515000)
	libcurand.so.9.0 => /usr/local/lib/python3.6/dist-packages/torch/lib/libcurand.so.9.0 (0x00007f37935b0000)
	libcusparse.so.9.0 => /usr/local/lib/python3.6/dist-packages/torch/lib/libcusparse.so.9.0 (0x00007f378fe45000)

```

**Comment by zakizhou** [comment]
I met the same problem when updating pytorch to 0.3, is there any solution?

**Comment by BernardoGO** [comment]
@zakizhou
I managed to "fix" the problem by changing the CUDA path and LD_LIBRARY_PATH. For some reason, instead of showing that it couldn't find a CUDA related library it just segfaults.

For me the problem was that Ubuntu really messes things up while installing Nvidia and CUDA. There was like four different folders for CUDA inside /use/local/.

Unfortunately, fixing for PyTorch broke Tensorflow in a way I couldn't even reinstall it.
Long story short, I've removed Ubuntu and installed Manjaro. Installed both tensorflow and PyTorch from AUR. Got everything working nice and shine within minutes.

**Comment by xiangyu1989** [comment]
@BernardoGO , 

could you give us some more details on how to change the CUAD path to the LD_LIBRARY_PATH

**Comment by sytrus-in-github** [comment]
I am using Ubuntu 16.04 with nvidia-driver 384.90 (and cuda 8) and pytorch 0.3.0.post4. Installed via pip.

Using e.g. "NVIDIA X Server Settings" I am able to switch between the integrated intel gpu and the nvidia one. I remarked that when I use the nvidia one everything works fine. However when I switch to the intel one I get the same segfault when importing pytorch. (should hope that it works with cpu only instead.) Hope this could somehow help your debugging process.

I have also used pytorch 0.2 before (also installed via pip) and didn't recall any problem like this. (or maybe I was lucky and never used pytorch under intel gpu before?)

**Comment by dhaalves** [comment]
Any news on that? i getting this error too in ubuntu 17.10 cuda 9.1

**Comment by aseembits93** [comment]
Append '/usr/lib/nvidia-384' or whichever nvidia driver version is being used to the LD_LIBRARY_PATH Environment variable. This worked out for me! Here is a reference - https://stackoverflow.com/questions/42678439/importerror-libnvidia-fatbinaryloader-so-375-39-cannot-open-shared-object-file

**Comment by soumith** [comment]
closing as duplicate of https://github.com/pytorch/pytorch/issues/4101

**Comment by soichih** [comment]
@soumith 
What do you mean by "closing as duplicate of #4101". This *is* the #4101 no?

@BernardoGO
>  For some reason, instead of showing that it couldn't find a CUDA related library it just segfaults.

I can't figure out why pytorch is crashing.. How did you find out why file pytorch is failing to find?

**Comment by soumith** [comment]
@soichih I must've entered the duplicate issue number wrong. This particular segfault has been fixed with `0.4.0` and `0.4.1` releases. The underlying issue was wrt statically linking libstdc++ and we no longer do that.

It's likely you have a different issue than this one.

**Comment by soichih** [comment]
@soumith Thanks for the info. I've upgraded to 0.4.1, and now `import torch` doesn't crash anymore, but now torch.cuda.is_available() is returning false.

```
>>> print(torch.cuda.is_available())
False
```

What do I need to do make CUDA available?

**Comment by soumith** [comment]
@soichih only a working NVIDIA driver should be needed. Does the command `nvidia-smi` work in the terminal?

**Comment by soichih** [comment]
@soumith On the host machine, yes.

```
$ nvidia-smi
Mon Aug 27 13:11:37 2018       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.77                 Driver Version: 390.77                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 1070    On   | 00000000:09:00.0  On |                  N/A |
|  0%   29C    P8    12W / 166W |    166MiB /  8116MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1521      G   /usr/lib/xorg/Xorg                           135MiB |
|    0     11731      G   compiz                                        28MiB |
+-----------------------------------------------------------------------------+

```

However, I am running our App (TractSeg) via singularity, and inside the container, it only has access to various library paths.

```
/usr/local/nvidia-390
/usr/local/cuda-8.0/lib64
```

I am guessing I have to somehow tell pytorch where to look for these libraries manually?

**Comment by soumith** [comment]
@soichih singularity probably does not have proper device paths setup? `nvidia-docker` sets up things correctly so that `/dev/nv*` is exposed correctly.

**Comment by hongyx11** [comment]
I met segment fault problem but It is the caffe2.so's problem, it works after I delete the build directory of caffe2.

**Comment by my-hello-world** [comment]
@soumith hi
i tried

> `**scratches head**`
> 
> how about:
> 
> ```
> gdb python
> r -c "import torch; print(torch)"
> ```
got
`Starting program: /home/vivian/anaconda3/envs/detect_and_track/bin/python -c "import torch"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff3417700 (LWP 15417)]
[New Thread 0x7ffff2c16700 (LWP 15418)]
[New Thread 0x7ffff0415700 (LWP 15419)]
[New Thread 0x7fffedc14700 (LWP 15420)]
[New Thread 0x7fffeb413700 (LWP 15421)]
[New Thread 0x7fffe8c12700 (LWP 15422)]
[New Thread 0x7fffe6411700 (LWP 15423)]
[New Thread 0x7fffe3c10700 (LWP 15424)]
[New Thread 0x7fffe140f700 (LWP 15425)]
[New Thread 0x7fffdec0e700 (LWP 15426)]
[New Thread 0x7fffdc40d700 (LWP 15427)]
[New Thread 0x7fffd9c0c700 (LWP 15428)]
[New Thread 0x7fffd740b700 (LWP 15429)]
[New Thread 0x7fffd4c0a700 (LWP 15430)]
[New Thread 0x7fffd2409700 (LWP 15431)]
[New Thread 0x7fffcfc08700 (LWP 15432)]
[New Thread 0x7fffcd407700 (LWP 15433)]
[New Thread 0x7fffcac06700 (LWP 15434)]
[New Thread 0x7fffc8405700 (LWP 15435)]
[New Thread 0x7fffc5c04700 (LWP 15436)]
[New Thread 0x7fffc3403700 (LWP 15437)]
[New Thread 0x7fffc0c02700 (LWP 15438)]
[New Thread 0x7fffbe401700 (LWP 15439)]
[Thread 0x7fffbe401700 (LWP 15439) exited]
[Thread 0x7fffcfc08700 (LWP 15432) exited]
[Thread 0x7fffc3403700 (LWP 15437) exited]
[Thread 0x7ffff0415700 (LWP 15419) exited]
[Thread 0x7fffe6411700 (LWP 15423) exited]
[Thread 0x7fffc5c04700 (LWP 15436) exited]
[Thread 0x7fffd9c0c700 (LWP 15428) exited]
[Thread 0x7fffe140f700 (LWP 15425) exited]
[Thread 0x7fffc0c02700 (LWP 15438) exited]
[Thread 0x7fffedc14700 (LWP 15420) exited]
[Thread 0x7ffff3417700 (LWP 15417) exited]
[Thread 0x7fffc8405700 (LWP 15435) exited]
[Thread 0x7fffcac06700 (LWP 15434) exited]
[Thread 0x7fffe8c12700 (LWP 15422) exited]
[Thread 0x7ffff2c16700 (LWP 15418) exited]
[Thread 0x7fffd2409700 (LWP 15431) exited]
[Thread 0x7fffd4c0a700 (LWP 15430) exited]
[Thread 0x7fffcd407700 (LWP 15433) exited]
[Thread 0x7fffdc40d700 (LWP 15427) exited]
[Thread 0x7fffd740b700 (LWP 15429) exited]
[Thread 0x7fffeb413700 (LWP 15421) exited]
[Thread 0x7fffdec0e700 (LWP 15426) exited]
[Thread 0x7fffe3c10700 (LWP 15424) exited]
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "torch/__init__.py", line 84, in <module>
    from torch._C import *
ImportError: No module named _C
[Inferior 1 (process 15413) exited with code 01]
`
please

**Comment by Suraj2794** [comment]
I started having this issue when i removed gpu form my machine. 
Did sudo apt-get purge nvidia* to get pytorch working.
